pkg/cloudprovider/providers/openstack/openstack.go

```diff
@@ -886,43 +932,30 @@ func (os *OpenStack) getVolume(diskName string) (volumes.Volume, error) {
 	return volume, err
 }
 
-func (os *OpenStack) getComputeIDbyHostname(cClient *gophercloud.ServiceClient) (string, error) {
+// getLocalComputeID tries to find ID of the OpenStack instance where
+// this code runs.
+func (os *OpenStack) getLocalComputeID(cClient *gophercloud.ServiceClient) (string, error) {
+	// try to read /var/lib/cloud/data/instance-id
+	const instanceIDFile = "/var/lib/cloud/data/instance-id"
+	idBytes, err := ioutil.ReadFile(instanceIDFile)
+	if err == nil {
+		id := string(idBytes)
+		id = strings.TrimSpace(id)
+		glog.V(5).Infof("Got instance id from %s: %s", instanceIDFile, id)
+		return id, nil
+	}
+	glog.V(4).Infof("Reading %s failed: %v, trying hostname", instanceIDFile, err)
 
+	// try to find ID from hostname
 	hostname, err := ossys.Hostname()
-
 	if err != nil {
 		glog.Errorf("Failed to get Minion's hostname: %v", err)
 		return "", err
 	}
-
-	i, ok := os.Instances()
-	if !ok {
-		glog.Errorf("Unable to get instances")
-		return "", errors.New("Unable to get instances")
-	}
-
-	srvs, err := i.List(".")
+	srv, err := getServerByHostName(cClient, hostname)
 	if err != nil {
-		glog.Errorf("Failed to list servers: %v", err)
+		glog.Errorf("Failed to find server for hostname %s: %v", hostname, err)
 		return "", err
 	}
-
-	if len(srvs) == 0 {
-		glog.Errorf("Found no servers in the region")
-		return "", errors.New("Found no servers in the region")
-	}
-	glog.V(4).Infof("Found servers: %v", srvs)
-
-	for _, srvname := range srvs {
-		server, err := getServerByName(cClient, srvname)
-		if err != nil {
-			return "", err
-		} else {
-			if (server.Metadata["hostname"] != nil && server.Metadata["hostname"] == hostname) || (len(server.Name) > 0 && server.Name == hostname) {
```

Comment:
No the server name is not guaranteed to match the hostname *at all* - which is why I'm trying to move this change to using the appropriate metadata apis instead..